### PR TITLE
BREAKING(bundler): Change Hook API to get_import_meta_props()

### PR DESCRIPTION
**Pull request by nayeemrmn**
Addresses https://github.com/denoland/deno/issues/8088.

### CODE REVIEW

File: spack/tests/pass/pr-1105/example-7/output/entry.js [hunk 1]
```diff
@@ -1,4 +1,16 @@
...
+}.url;
 console.log(isMain, modUrl);
-console.log(import.meta.main, "$DIR/tests/pass/pr-1105/example-7/input/entry.js");
+console.log(({
+    url: "$DIR/tests/pass/pr-1105/example-7/input/entry.js",
+    main: import.meta.main
+}).main, ({
+    url: "$DIR/tests/pass/pr-1105/example-7/input/entry.js",
+    main: import.meta.main
+}).url);
```
Comment (nayeemrmn) [inline spack/tests/pass/pr-1105/example-7/output/entry.js:16]: WIP: I'd like this to emit like this:
```js
const import_meta_1 = {
    url: "$DIR/tests/pass/pr-1105/example-7/input/f.js",
    main: false
};
const isMain = import_meta_1.main;
const modUrl = import_meta_1.url;
console.log(isMain, modUrl);
const import_meta_2 = {
    url: "$DIR/tests/pass/pr-1105/example-7/input/entry.js",
    main: import.meta.main
};
console.log(import_meta_2.main, import_meta_2.url);
```
as I originally suggested in https://github.com/denoland/deno/issues/8088. I left a TODO in `merge.rs`. Since it's a bit more involved, I'd like to leave it for another PR.

File: bundler/src/bundler/chunk/merge.rs [hunk 2]
```diff
@@ -626,7 +625,6 @@ impl VisitMut for DefaultRenamer {
 struct ImportMetaHandler<'a, 'b> {
     file: &'a FileName,
     hook: &'a Box<dyn 'b + Hook>,
-    is_entry: bool,
```
Comment (kdy1) [inline bundler/src/bundler/chunk/merge.rs:629]: I think this field can help deducing if `import.meta.main` is false.

File: bundler/src/hook.rs [hunk 3]
```diff
@@ -1,9 +1,12 @@
 use anyhow::Error;
 use swc_common::{FileName, Span};
-use swc_ecma_ast::Expr;
+use swc_ecma_ast::KeyValueProp;
 
 /// Note: As this is rarely used, it's recommended to pass it as a trait object.
 pub trait Hook: swc_common::sync::Sync + swc_common::sync::Send {
-    /// Return [None] to preserve `import.meta.url`
-    fn get_import_meta_url(&self, span: Span, file: &FileName) -> Result<Option<Expr>, Error>;
```
Comment (kdy1) [inline bundler/src/hook.rs:8]: I think the api should provide a way to determine if it's main.

**Review by kdy1 — COMMENTED** [review]
Thanks a lot!
I forgot about this.

File: bundler/src/hook.rs (hunk 3 above)
Comment (nayeemrmn) [reply bundler/src/hook.rs:8]: What do you think about this structure?
```rust
use anyhow::Error;
use swc_common::{FileName, Span};
use swc_ecma_ast::KeyValueProp;

#[non_exhaustive]
pub struct ModuleRecord {
  file_name: FileName,
  is_entry: bool,
}

pub trait Hook: swc_common::sync::Sync + swc_common::sync::Send {
    fn get_import_meta_props(
        &self,
        span: Span,
        module_record: &ModuleRecord,
    ) -> Result<Vec<KeyValueProp>, Error>;
}
```

Comment (nayeemrmn) [reply bundler/src/hook.rs:8]: Done.

Comment (kdy1) [reply bundler/src/hook.rs:8]: Seems great! Thanks.

**Review by kdy1 — APPROVED** [review]
Thanks a lot!
I'll merge this after handling https://github.com/swc-project/swc/pull/1196

**Comment by nayeemrmn** [comment]
I realised the visitor was only catching `import.meta` in member expressions, pushed a fix.

File: bundler/src/bundler/chunk/merge.rs [hunk 4]
```diff
@@ -637,74 +637,46 @@ struct ImportMetaHandler<'a, 'b> {
...
+            }) => match self.hook.get_import_meta_props(
+                *span,
+                &ModuleRecord {
+                    file_name: self.file.to_owned(),
+                    is_entry: self.is_entry,
+                },
+            ) {
+                // TODO(nayeemrmn): This substitutes any `import.meta` reference with a complete
+                // re-instantiated object. Support mutating `import.meta` using pre-declared a
+                // variable like `import_meta_1`.
```
Comment (nayeemrmn) [inline bundler/src/bundler/chunk/merge.rs:666]: @kdy1 Regarding this, is there an easy way of getting a variable name not used as a local var in the module? If not I think it needs an extra pass, which I can also use to see if `import.meta` is ever used.

Comment (kdy1) [reply bundler/src/bundler/chunk/merge.rs:666]: You can use `swc_ecma_utils::private_ident!`.
`hygiene` pass will rename it to an uniqueidentifier.
